# Задание SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать. 

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта

# ЗАДАНИЯ

* Посчитать, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитать количество обзоров и среднюю оценку;
* Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
* Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
* Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

# Ипорт библиотек

In [1]:
import pandas as pd
import sqlalchemy as sa
# устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)

In [2]:
def dataset_info(dataset):
    display(dataset.head(5))
    display('Количество пустых ячеек')
    display(dataset.isnull().sum())
    display(dataset.info())
    display('Количество явных дубликатов в датасете', dataset.duplicated().sum())

In [3]:
tables=['books','authors','publishers','ratings','reviews']
for i in tables:
    display(f'Анализ таблицы {i}')
    query = '''SELECT * FROM {}'''.format(i)
    get_sql_data(query)
    data = pd.io.sql.read_sql(query, con = engine)
    dataset_info(data)

'Анализ таблицы books'

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

'Количество пустых ячеек'

book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

'Количество явных дубликатов в датасете'

0

'Анализ таблицы authors'

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

'Количество пустых ячеек'

author_id    0
author       0
dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

'Количество явных дубликатов в датасете'

0

'Анализ таблицы publishers'

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

'Количество пустых ячеек'

publisher_id    0
publisher       0
dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

'Количество явных дубликатов в датасете'

0

'Анализ таблицы ratings'

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

'Количество пустых ячеек'

rating_id    0
book_id      0
username     0
rating       0
dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

'Количество явных дубликатов в датасете'

0

'Анализ таблицы reviews'

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

'Количество пустых ячеек'

review_id    0
book_id      0
username     0
text         0
dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

'Количество явных дубликатов в датасете'

0

# ВЫВОД

### По результатам загрузки в таблицах 'books','authors','publishers','ratings','reviews' отсутствуют дубликаты, пропуски, некорректно названные столбцы

# Количество вышедших книг после 1 января 2000 года

In [4]:
query = '''SELECT COUNT(books.title) 
FROM books 
WHERE CAST(books.publication_date AS TIMESTAMP)>=CAST('2000-01-01' AS TIMESTAMP) LIMIT 5'''
get_sql_data(query)

count
0    821

# Количество обзоров и средня оценка книг

In [5]:
query = '''SELECT b.title, COUNT(DISTINCT(r.review_id)), AVG(rt.rating)
FROM books AS b
LEFT JOIN reviews AS r ON b.book_id=r.book_id
LEFT JOIN ratings AS rt ON b.book_id=rt.book_id
GROUP BY b.book_id
ORDER BY AVG(rt.rating) DESC'''
get_sql_data(query)

title  count   avg
0        Arrows of the Queen (Heralds of Valdemar  #1)      2  5.00
1    The Walking Dead  Book One (The Walking Dead #...      2  5.00
2                                      Light in August      2  5.00
3    Wherever You Go  There You Are: Mindfulness Me...      2  5.00
4    Captivating: Unveiling the Mystery of a Woman'...      2  5.00
..                                                 ...    ...   ...
995  The World Is Flat: A Brief History of the Twen...      3  2.25
996                  His Excellency: George Washington      2  2.00
997                                      Drowning Ruth      3  2.00
998                                              Junky      2  2.00
999                               Harvesting the Heart      2  1.50

[1000 rows x 3 columns]

# Издательство, которое выпустило наибольшее число книг толще 50 страниц 

In [6]:
query = '''SELECT p.publisher, COUNT(DISTINCT(b.book_id))
FROM books AS b
LEFT JOIN publishers AS p ON b.publisher_id=p.publisher_id
WHERE b.num_pages>50
GROUP BY p.publisher
ORDER BY COUNT(DISTINCT(b.book_id)) DESC
LIMIT 1'''
get_sql_data(query)

publisher  count
0  Penguin Books     42

# Автор с самой высокой средней оценкой книг* 


*(учитываются книги с 50 и более оценками)

In [7]:
query = '''SELECT a.author, AVG(rt.rating)
FROM authors AS a
JOIN books AS b ON a.author_id=b.author_id
JOIN reviews AS r ON r.book_id=r.book_id
JOIN ratings AS rt ON b.book_id=rt.book_id
WHERE b.book_id IN (SELECT b.book_id FROM books AS b
LEFT JOIN ratings AS r ON b.book_id=r.book_id
GROUP BY b.book_id
HAVING COUNT(r.rating)>=50)
GROUP BY  a.author_id
ORDER BY AVG(rt.rating) DESC
LIMIT 1'''
get_sql_data(query)

author       avg
0  J.K. Rowling/Mary GrandPré  4.287097

# Среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [8]:
query = '''
SELECT AVG(av.count)
FROM (SELECT r.username, COUNT(r.review_id)
      FROM reviews AS r
      WHERE r.username IN (SELECT rt.username 
                           FROM ratings AS rt
                           GROUP BY rt.username 
                           HAVING COUNT(rt.rating)>48)
      GROUP BY r.username) AS av
'''
get_sql_data(query)

avg
0  24.0

### ВЫВОД:
По результатам исследования получены следующие данные:
* Количество вышедших книг после 1 января 2000 года-821
* Количество обзоров и средня оценка книг  
* Издательство, которое выпустило наибольшее число книг толще 50 страниц- Penguin Books 42
* Автор с самой высокой средней оценкой книг (учитываются книги с 50 и более оценками)-J.K. Rowling/Mary GrandPré 4.287097
* Среднее количество обзоров от пользователей, которые поставили больше 48 оценок-22.882353